packages/frontend-app-api/src/wiring/createApp.test.tsx

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  createExtension,
+  createExtensionDataRef,
+  createExtensionInput,
+  createPageExtension,
+  createPlugin,
+} from '@backstage/frontend-plugin-api';
+import { createInstances } from './createApp';
+
+import { MockConfigApi } from '@backstage/test-utils';
+import React from 'react';
+import { createRouteRef } from '@backstage/core-plugin-api';
+
+describe('createInstances', () => {
+  it('throws an error when a root extension is parametrized', () => {
+    const config = new MockConfigApi({
+      app: {
+        extensions: [
+          {
+            root: {
+              at: '',
+            },
+          },
+        ],
+      },
+    });
+    const plugins = [
+      createPlugin({
+        id: 'plugin',
+        extensions: [],
+      }),
+    ];
+    expect(() => createInstances({ config, plugins })).toThrow(
+      "A 'root' extension configuration was detected, but the root extension is not configurable",
+    );
+  });
+
+  it('throws an error when a root extension is overridden', () => {
+    const config = new MockConfigApi({});
+    const plugins = [
+      createPlugin({
+        id: 'plugin',
+        extensions: [
+          createExtension({
+            id: 'root',
+            at: 'core.routes/route',
+            inputs: {},
+            output: {},
+            factory() {},
+          }),
+        ],
+      }),
+    ];
+    expect(() => createInstances({ config, plugins })).toThrow(
+      "The following plugin(s) are overriding the 'root' extension which is forbidden: plugin",
+    );
+  });
+
+  describe('should deduplicated plugins', () => {
+    it('in an immediate dependency level (e.g., A -> A)', () => {
+      const config = new MockConfigApi({});
+
+      const addonExtensionData =
+        createExtensionDataRef<JSX.Element>('plugin.page.addon');
+
+      const CyclicalA = createExtension({
+        id: 'A', // Same id as ancestor A
+        at: 'A/addons',
+        inputs: {},
+        output: {
+          element: addonExtensionData,
+        },
+        factory({ bind }) {
+          bind({
+            element: <div key="cyclical-a">Cyclical A</div>,
+          });
+        },
+      });
+
+      const A = createPageExtension({
```

Comment:
Could just make this a test for two simple page extensions instead right? without inputs. We're just making sure that plugins can't provide extensions with conflicting IDs. Might be good to make sure it detects conflicts across plugins too, probably even conflict between core extensions and plugins.

Reply:
Done 🙌🏻 : https://github.com/backstage/backstage/pull/20032/commits/c71da2a7907c53e6d147b76925ee0f1341406669